articles/remapping.md

```diff
@@ -0,0 +1,119 @@
+
+#### Pre-FQN expansion remapping
+Say the source code for a node use two names `cat` and `/ns/cat`.
+If the node runs in namespace `ns`, the FQN of both topics will be identical: `/ns/cat`.
+No rule could remap these names individually once they have been expanded.
+A rule applied Pre-FQN expansion could remap just one of them.
+
+#### Requirements for remapping capabilities
+##### 1. Namespace Replacement
+These are rules that only apply to the namespace part of the name.
+
+**A user can remap the first namespace token without replacing all namespace tokens with the same name.**
+
+*Example*
+Node uses names `/big/cat`, `/big/red/dog`, `/red/big/cat`.
+It must be possible to make a rule that replaces "big" with "small" such that the final names are `/small/cat`, `/small/red/dog`, `/red/big/cat`
+
+**A user can replace all matching namespace tokens without changing the basename.**
+
+*Example*
+Node uses names `/big/cat`, `/red/big/dog`, `/dog/big`.
+It must be possible to make a rule that replaces "big" with "small" such that the final names are `/small/cat`, `/red/small/dog`, `/dog/big`
+
+**A user can replace multiple matching namespace tokens only if whole namespaces are matched.**
+
+*Example*
+Node uses names `/big/red/dog`, `/very_big/red/dog`.
+It must be possible to make a rule that replaces "big/red" with "small" such that the final names are `/small/dog`, `/very_big/red/dog`
+
+##### 2. Basename Replacement
+**A user can replace a full basename token with another basename token**
+
+*Example*
+Node uses names `/big/cat`, `/big/cat/paws`.
+It must be possible to make a rule that replaces "cat" with "dog" such that the final names are `/big/dog`, `/big/cat/paws`
+
+##### 3. Exact Replacement
+**A user can replace a full name with another name**
+
+*Example*
+Node uses names `/big/cat`, `/big/cat/paws`.
+It must be possible to make a rule that replaces "/big/cat" with "/bear" such that the final names are `/bear`, `/big/cat/paws`
+
+##### 4. Arbitrary Substring Replacement ***(maybe)***
+**A user can replace any substring with another string**
+
+*Examples*
+Node uses names `/big/red/dog`, `/very_big/red/dog`.
+It must be possible to make a rule that replaces "big/red" with "small" such that the final names are `/small/dog`, `/very_small/dog`
+
+Node uses names `/big/dog`, `/dog/big`.
+It must be possible to make a rule that replaces "dog" with "cat" such that the final names are `/big/cat`, `/cat/big`
+
+Node uses names `/big/red/dog`, `/small/cat`.
+Node is run in namespace `ns`.
+It must be possible to make a rule that replaces "/" with "_" such that the final names are `/ns/_big_red_dog`, `/ns/_small_cat`
+
+#### Static Remapping
+Static remapping is the ability to provide a node with remap rules prior to launching it.
+A node stores all remapping rules for the duration of it's life.
+All names have remapping rules applied before they are used.
+This means the DDS interface, and maybe all of `rmw`, is ignorant of remapping.
```

Comment:
How about applying rules  in the order they're specified? In that case, a name `foo` would become `bar` after the first rule, and the second rule would change nothing because there are no `foo` names to act on.

I think starting with that logic makes dynamic remapping easier later on, because dynamic rules should chain to let a user remap the names they see at runtime.